### ISSUE TITLE
EF Core 9 support

### ISSUE THREAD

**Issue by lauxjpn** [PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#1841]
We will use this issue to track the progress on supporting EF Core 9, including its previews.

**Comment by lauxjpn** [comment]
Prerelease [9.0.0-preview.1](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/9.0.0-preview.1) (compatible with EF Core `9.0.0-preview.1.24081.2`) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/9.0.0-preview.1).

**Comment by rodolforeinaperficient** [comment]
Hi @lauxjpn, do you have plans to update to 9.0.0-preview.5?

**Comment by ysmoradi** [comment]
> Hi @lauxjpn, do you have any plans to update to 9.0.0-preview.5?

The same for .NET 9 preview7 ?

**Comment by dajma00** [comment]
I think we need a release for .NET 9, RC1.

**Comment by lauxjpn** [comment]
An EF Core 9 RC release will be out in about 2-3 weeks.

**Comment by dajma00** [comment]
Anyone else getting this both with Pomelo RC1 and RC2, .NET9 RC2:
![image](https://github.com/user-attachments/assets/a785e4f3-30f0-45c2-8da3-58ff8408fb71)

**Comment by glennawatson** [comment]
> Anyone else getting this both with Pomelo RC1 and RC2, .NET9 RC2: ![image](https://private-user-images.githubusercontent.com/38800605/375650452-a785e4f3-30f0-45c2-8da3-58ff8408fb71.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4MzA0ODIsIm5iZiI6MTcyOTgzMDE4MiwicGF0aCI6Ii8zODgwMDYwNS8zNzU2NTA0NTItYTc4NWU0ZjMtMzBmMC00NWMyLThkYTMtNThmZjg0MDhmYjcxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDA0MjMwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhY2NjYjE0YTBlYTFmMmJhNWMzMDljOWI2Mzg5YzZjZGFlZTdiMDZlZjQyNGM0YzQwMTZlODNiNTFhMDdjMDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.FC75ZKJKK3saIUF_rgM4uZmeerOoFsAGnvboST4q7_U)

Yeah, seems to be more members were added in the RC vs the preview in entity framework itself.

**Comment by bkoelman** [comment]
@lauxjpn Can we please get a NuGet release that's compatible with RC2 soon? I'd like to run our 9000+ test suite against it. I don't mind using the pomelo-nightly feed.

**Comment by raffidahmad** [comment]
Hi, what's the timeline for a RC2 compatible release?

**Comment by siriak** [comment]
Looking forward to .NET 9 version 👀

**Comment by bkoelman** [comment]
@bdovaz I think there *are* breaking changes in EF Core 9, in the pubternal types that providers need to implement. That's why suppressing the version range warning doesn't make it work.

**Comment by bdovaz** [comment]
@bkoelman I mean that “maybe not always” there are breaking changes that affect this library between major versions and that it is limiting by default its use

**Comment by lauxjpn** [comment]
I will try to get a compatible version out early next week (I am very much swamped this week, so earlier will just not be possible). While this prerelease will be compatible with EF Core 9, it will probably not contain a lot of the new EF Core 9 features yet.

---

> But because of the decision to use ranges in the EF Core dependency, we always depend on a new stable release specific for that version...

@bdovaz That is a deliberate decision (doesn't have to be stable though). Every major EF Core release contains breaking changes and is incompatible with provider releases that target any preceding major version.

**Comment by Emiliowilvca** [comment]
when is version 9.0.0 released?

**Comment by glennawatson** [comment]
> when is version 9.0.0 released?

He just replied above your message. He's hoping this week but swamped

**Comment by kuerhero** [comment]
When will efcore9.0 be supported?

**Comment by fairok** [comment]
> When will efcore9.0 be supported?

Maybe early in this week, please refer the above message.

**Comment by nicholasyin** [comment]
I personally don't expect the new version to support EF9 new features. Just making it compatible and working with EF9 is enough for me.

**Comment by zijianhuang** [comment]
I am pretty sure the author will deliver soon when getting a break.

**Comment by mapo80** [comment]
> I am pretty sure the author will deliver soon when getting a break.

I hope so!

**Comment by sveinl** [comment]
> > I am pretty sure the author will deliver soon when getting a break.
> 
> I hope so!

Same, it was pretty bad to go back to .NET 8 after working with .NET 9 for a while :)

**Comment by srad** [comment]
> > > I am pretty sure the author will deliver soon when getting a break.
> > 
> > 
> > I hope so!
> 
> Same, it was pretty bad to go back to .NET 8 after working with .NET 9 for a while. Still we should'nt demand anyting from voluntary work, just wait patiently and then say thanks! :)

I would donate time to work on it, but the documentation is non-existent on how to work on the codebase, it's too difficult to get into it. I took a look, pretty impossible.

**Comment by BartoGabriel** [comment]
> Don't know why Microsoft/Oracle/MariaDB don't give Pomelo authors any technical or finanical support, since there is only one EFCore package work with MySql/MariaDB.

There's a thread on Twitter where Microsoft team members are discussing issues preventing migration to .NET 9. Perhaps we could engage with them to explore how Microsoft could contribute to this project, as support for MySQL and MariaDB is essential for any ORM.

https://x.com/mjovanovictech/status/1859580143460626642

**Comment by lauxjpn** [comment]
The basic migration to EF Core 9.0 (making the provider and test suite compile) is now complete.
I will evaluate the state of the test suite and gather some statistics and will then open a corresponding PR (in about 24 hours).

**Comment by MPITech** [comment]
> The basic migration to EF Core 9.0 (making the provider and test suite compile again) is now complete. I will evaluate the state of the test suite and gather some statistics and will then open a corresponding PR (in about 24 hours).

You are fantastic.  Thank you!

**Comment by DLozanoNavas** [comment]
As a community, I believe we should come together to financially support this project. It plays a crucial role for many organizations and developers, and the dedication of the authors deserves recognition. Setting up channels like GitHub Sponsors or similar platforms could make it easier for us to contribute. Even small amounts can go a long way—at the very least, to show our appreciation and maybe even buy them some well-earned coffee. Their work is invaluable, and our support could help ensure its continued success.

**Comment by kampilan** [comment]
> As a community, I believe we should come together to financially support this project. It plays a crucial role for many organizations and developers, and the dedication of the authors deserves recognition. Setting up channels like GitHub Sponsors or similar platforms could make it easier for us to contribute. Even small amounts can go a long way—at the very least, to show our appreciation and maybe even buy them some well-earned coffee. Their work is invaluable, and our support could help ensure its continued success.

I have been saying this for years. I would be the first to pony up a few hundred $ each year. I have come to realize it is a bit insulting to Laurents et al. As if he is just sitting around goofing off because he is not getting paid. The only way this would work is if he wanted to quit his day job and do this full time instead. Maybe he likes what he does or makes a good salary. What this project needs it more people.  Not easy. Steep learning curve.  As stated above I will wait patiently and thankfully.

**Comment by DLozanoNavas** [comment]
> > As a community, I believe we should come together to financially support this project. It plays a crucial role for many organizations and developers, and the dedication of the authors deserves recognition. Setting up channels like GitHub Sponsors or similar platforms could make it easier for us to contribute. Even small amounts can go a long way—at the very least, to show our appreciation and maybe even buy them some well-earned coffee. Their work is invaluable, and our support could help ensure its continued success.
> 
> I have been saying this for years. I would be the first to pony up a few hundred $ each year. I have come to realize it is a bit insulting to Laurents et al. As if he is just sitting around goofing off because he is not getting paid. The only way this would work is if he wanted to quit his day job and do this full time instead. Maybe he likes what he does or makes a good salary. What this project needs it more people. Not easy. Steep learning curve. As stated above I will wait patiently and thankfully.

I appreciate your perspective, but there might be a misunderstanding about the intention behind my suggestion. The idea isn’t to pressure the authors or imply that they’re not doing enough. Quite the opposite—it’s about expressing gratitude and supporting them in a tangible way for the incredible value they provide to the community.

Of course, it’s entirely their choice how they want to approach funding or whether they even want to consider it. The suggestion is an additional way for us as a community to show our appreciation beyond patiently and thankfully waiting for their work. Their dedication is remarkable, and our small contributions could be a way to acknowledge that—not to replace their passion or day jobs, but to say, “We see and value what you do.”

**Comment by kampilan** [comment]
> > > As a community, I believe we should come together to financially support this project. It plays a crucial role for many organizations and developers, and the dedication of the authors deserves recognition. Setting up channels like GitHub Sponsors or similar platforms could make it easier for us to contribute. Even small amounts can go a long way—at the very least, to show our appreciation and maybe even buy them some well-earned coffee. Their work is invaluable, and our support could help ensure its continued success.
> > 
> > 
> > I have been saying this for years. I would be the first to pony up a few hundred $ each year. I have come to realize it is a bit insulting to Laurents et al. As if he is just sitting around goofing off because he is not getting paid. The only way this would work is if he wanted to quit his day job and do this full time instead. Maybe he likes what he does or makes a good salary. What this project needs it more people. Not easy. Steep learning curve. As stated above I will wait patiently and thankfully.
> 
> @kampilan I appreciate your perspective, but there might be a misunderstanding about the intention behind my suggestion. The idea isn’t to pressure the authors or imply that they’re not doing enough. Quite the opposite—it’s about expressing gratitude and supporting them in a tangible way for the incredible value they provide to the community.
> 
> Of course, it’s entirely their choice how they want to approach funding or whether they even want to consider it. The suggestion is an additional way for us as a community to show our appreciation beyond patiently and thankfully waiting for their work. Their dedication is remarkable, and our small contributions could be a way to acknowledge that—not to replace their passion or day jobs, but to say, “We see and value what you do.”

I probably did not word this as elegantly as I had hoped. I feel just like you do. The people that maintain this project deserve compensation and I am ready to pay. I was simply musing out loud as to why he resists these calls. Maybe he feels there aren't enough "buyers" to support full time work on the project. Maybe he thinks taking money changes the nature of his involvement. Maybe he really doesn't have the time no matter the price. Since I have no one to give my money to, I just wait patiently and thankfully.

**Comment by benpsnyder** [comment]
Please make it clear in the README how we can financially contribute. My company https://snyder.tech will certainly give a financial contribution. Thank you!

**Comment by CumpsD** [comment]
> The basic migration to EF Core 9.0 (making the provider and test suite compile again) is now complete. I will evaluate the state of the test suite and gather some statistics and will then open a corresponding PR (in about 24 hours).

Looking forward to `net9.0` support!

**Comment by mapo80** [comment]
> > The basic migration to EF Core 9.0 (making the provider and test suite compile again) is now complete. I will evaluate the state of the test suite and gather some statistics and will then open a corresponding PR (in about 24 hours).
> 
> Looking forward to `net9.0` support!

As everyone here! ;)

**Comment by m3nax** [comment]
I believe that an RC or Preview version with the updated dependencies is more than sufficient to allow us to start testing net9.0 with the new version of the framework, currently the problem with preview1 is that it has the dependency to the preview1.24081 version of the Microsoft.EntityFrameworkCore.Relational nuget inside it.

**Comment by mapo80** [comment]
> I believe that an RC or Preview version with the updated dependencies is more than sufficient to allow us to start testing net9.0 with the new version of the framework, currently the problem with preview1 is that it has the dependency to the preview1.24081 version of the Microsoft.EntityFrameworkCore.Relational nuget inside it.

It's not that simple! I've tried to use latest version of Microsoft.EntityFrameworkCore.Relational and there are lots of things to do

**Comment by m3nax** [comment]
> > I believe that an RC or Preview version with the updated dependencies is more than sufficient to allow us to start testing net9.0 with the new version of the framework, currently the problem with preview1 is that it has the dependency to the preview1.24081 version of the Microsoft.EntityFrameworkCore.Relational nuget inside it.
> 
> It's not that simple! I've tried to use latest version of Microsoft.EntityFrameworkCore.Relational and there are lots of things to do

Good to know

**Comment by JochemPalmsens** [comment]
> Hi, I've tried updated version of sdk building it from source.

Hi, Please read the opening post
> We will use this issue to track the progress on supporting EF Core 9, including its previews.

So please don't use this thread (that everybody is interested in progress updates is subscribed to) for some issue you encountered: [open a new issue](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/new/choose) (or just wait until there's an official release)

**Comment by ysiivan** [comment]
> System.TypeLoadException: 'Method 'get_LockReleaseBehavior' in type 'Pomelo.EntityFrameworkCore.MySql.Migrations.Internal.MySqlHistoryRepository' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=9.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.'

**Comment by linkdotnet** [comment]
> `System.TypeLoadException: 'Method 'get_LockReleaseBehavior' in type 'Pomelo.EntityFrameworkCore.MySql.Migrations.Internal.MySqlHistoryRepository' from assembly 'Pomelo.EntityFrameworkCore.MySql, Version=9.0.0.0, Culture=neutral, PublicKeyToken=2cc498582444921b' does not have an implementation.'`

You found this thread but didn’t care to read any of the messages?

**Comment by rgwsk65** [comment]
Do we know when it will be compatible with .NET 9? I'm not rushing, just asking 😊

**Comment by glennawatson** [comment]
> Do we know when it will be compatible with .NET 9? I'm not rushing, just asking 😊

There seems to be a testing version from above. There's already a fair bit of discussion above for more context.

**Comment by fairok** [comment]
> Do we know when it will be compatible with .NET 9? I'm not rushing, just asking 😊

We should patiently wait for the author's effort. According to the records of Github Actions, the MySQL test cases have passed, but Mariadb has not yet. But I guess this is also coming soon.

**Comment by chendjouCaleb** [comment]
Build successful ! Thank to author for his hard working.

**Comment by VIOVNL** [comment]
Almost there, just a NuGet authorization error left to fix. 🥳

**Comment by lauxjpn** [comment]
Alright, we have a _prerelease_ in our [nightly build feeds](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql?tab=readme-ov-file#nightly-builds) that is compatible with EF Core 9.0.x and passes all 31K existing tests for all (officially) supported MySQL and MariaDB versions.

There is an edge case that we handle now differently than before, and that will need more work before the GA release. I will open an issue about it and link it to this one and the PR above shortly.

**Comment by VIOVNL** [comment]
> Alright, we have a _prerelease_ in our [nightly build feeds](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql?tab=readme-ov-file#nightly-builds) that is compatible with EF Core 9.0.x and passes all 31K existing tests for all (officially) supported MySQL and MariaDB versions.
> 
> There is an [edge case](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1948/commits/c5bb93426f0ae78f733df34e28e26cdae841ec9a) that we handle now differently than before, and that will need more work before the GA release. I will open an issue about it and link it to this one and the PR above shortly.
> 
> Unless there are some serious issues with the _prerelease_, we will publish it on nuget.org in 2-3 days as `9.0.0-alpha.1`.
> 
> Thank you everybody for your patience!

Thank you for the work @lauxjpn 💪

For anyone wondering it's this one:

```
Install-Package Pomelo.EntityFrameworkCore.MySql -Version 9.0.0-alpha.1.ci.20241207210245 -Source https://www.myget.org/F/pomelo/api/v3/index.json 
```

**Comment by lauxjpn** [comment]
I copied the posts above, about ideas to financially support the project, to the new issue #1950, so we can discuss the subject independently of this issue here.

**Comment by lauxjpn** [comment]
> [...]
> Word of note for everyone: Visual Studio sorts prerelease releases by type. So it thinks that 9.0.0-preview.2.ci.20240426151710 is the latest prerelease version. You need to scroll down and look for 9.0.0-alpha.1.ci.20241207210245 in order to update to the latest version.

As pointed out by @MarkCiliaVincenti in https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1944#issuecomment-2525549436_, there is a discoverability issue with our traditional naming strategy (if we already released a _preview_ before).

I opened #1952 about it.

TLDR: We might change the name of the upcoming prerelease from `9.0.0-alpha.1` to something similar to `9.0.0-preview.alpha.1`, so it gets sorted on top of the prerelease list in IDEs etc.

**Comment by coachnrm** [comment]
I try to follow @lauxjpn by copying 
<PackageReference Include="Pomelo.EntityFrameworkCore.MySql" Version="9.0.0-alpha.1.ci.20241207210245" />
and dotnet restore and scaffold. It works. Thank you.

**Comment by NgLoader** [comment]
@dependabot recreate

**Comment by kanadaj** [comment]
@lauxjpn EF Core 9.0 supports Parameterized Collections, but Pomelo seems to throw on trying to translate it. It requires setting `TranslateParameterizedCollectionsToConstants` in the options for it to work at all. Might be best to set it by default to make upgrades smoother until it's fixed

**Comment by lauxjpn** [comment]
@kanadaj Please open an issue about it, so we can discuss it independently. This issue here is about tracking the overall progress to release Pomelo 9.0 GA. Thanks!

**Comment by tietaren** [comment]
@lauxjpn Thank you very much for your hard work! May I ask when the Pomelo 9.0 GA is expected to be released?

**Comment by lauxjpn** [comment]
@tietaren The GA version is likely to be released next month (January 2025). I'll update our roadmap when we publish our `9.0.0-preview.2.efcore.9.0.0` release (formerly known as `9.0.0-alpha.1`) this week.

**Comment by anoori-opus** [comment]
I installed the version "9.0.0-preview.2.ci.20240426151710" from DevOps repo and when I run my app, it gives me this error: Could not load type 'Microsoft.EntityFrameworkCore.TypeBaseExtensions' from assembly 'Microsoft.EntityFrameworkCore, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.

I understand this is not GA version but didn't know if I'm doing something wrong or not. Is this expected?

**Comment by CardenInsurance** [comment]
> I installed the version "9.0.0-preview.2.ci.20240426151710" from DevOps repo and when I run my app, it gives me this error: Could not load type 'Microsoft.EntityFrameworkCore.TypeBaseExtensions' from assembly 'Microsoft.EntityFrameworkCore, Version=9.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'.
> 
> I understand this is not GA version but didn't know if I'm doing something wrong or not. Is this expected?

20240426151710 is not the latest version available.  Look for 20241207210245 or newer.

**Comment by anoori-opus** [comment]
Thank you. That worked. I don't know why it didn't show up on the feed in the GUI so I installed it using command line.

**Comment by lauxjpn** [comment]
Prerelease [9.0.0-preview.2.efcore.9.0.0](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/9.0.0-preview.2.efcore.9.0.0) (compatible with EF Core `9.0.x`) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/9.0.0-preview.2.efcore.9.0.0).

**Comment by ErikEJ** [comment]
@lauxjpn Amazing  - I will get it added to EF Core Power Tools ASAP!

**Comment by iccfish** [comment]
With `9.0.0-preview.2.efcore.9.0.0` the following query can not work anymore:

```
var uids = new ulong[0]{};
var users = await this.dbContext.Users.Where(s=>uids.Contains(s.Uid)).ToListAsync();
```
The codes above works ok with 8.0.11.
Exception was:

```
      An unhandled exception has occurred while executing the request.
      System.InvalidOperationException: The LINQ expression '@__uids_0' could not be translated. Additional information: Primitive collections support has not been enabled. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
         at Pomelo.EntityFrameworkCore.MySql.Query.Internal.MySqlQueryableMethodTranslatingExpressionVisitor.TranslatePrimitiveCollection(SqlExpression sqlExpression, IProperty property, String tableAlias)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateParameterQueryRoot(ParameterQueryRootExpression parameterQueryRootExpression)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.TranslateSubquery(Expression expression)
         at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.<VisitMethodCall>g__TranslateAsSubquery|33_0(Expression expression, <>c__DisplayClass33_0&)
         at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.CallBaseVisitMethodCall(MethodCallExpression methodCallExpression)
         at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression, Boolean applyDefaultTypeMapping)
         at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.Translate(Expression expression, Boolean applyDefaultTypeMapping)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression, Boolean applyDefaultTypeMapping)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
         at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
         at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression[TResult](Expression query)
         at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
         at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass11_0`1.<ExecuteCore>b__0()
         at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteCore[TResult](Expression query, Boolean async, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
         at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
         at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
         at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
```

**Comment by miroslavpopovic** [comment]
> With `9.0.0-preview.2.efcore.9.0.0` the following query can not work anymore:
> 
> ```
> var uids = new ulong[0]{};
> var users = await this.dbContext.Users.Where(s=>uids.Contains(s.Uid)).ToListAsync();
> ```
> 
> The codes above works ok with 8.0.11. Exception was:
> 
> ```
>       An unhandled exception has occurred while executing the request.
>       System.InvalidOperationException: The LINQ expression '@__uids_0' could not be translated. Additional information: Primitive collections support has not been enabled. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
>          at Pomelo.EntityFrameworkCore.MySql.Query.Internal.MySqlQueryableMethodTranslatingExpressionVisitor.TranslatePrimitiveCollection(SqlExpression sqlExpression, IProperty property, String tableAlias)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateParameterQueryRoot(ParameterQueryRootExpression parameterQueryRootExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.TranslateSubquery(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.<VisitMethodCall>g__TranslateAsSubquery|33_0(Expression expression, <>c__DisplayClass33_0&)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.CallBaseVisitMethodCall(MethodCallExpression methodCallExpression)
>          at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.Translate(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression[TResult](Expression query)
>          at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
>          at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass11_0`1.<ExecuteCore>b__0()
>          at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteCore[TResult](Expression query, Boolean async, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
>          at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
>          at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
> ```

The same thing happens to me too.

**Comment by raffidahmad** [comment]
yup have noticed the same, before i thought it was because pomelo was on preview 1

**Comment by devnullue** [comment]
> With `9.0.0-preview.2.efcore.9.0.0` the following query can not work anymore:
> 
> ```
> var uids = new ulong[0]{};
> var users = await this.dbContext.Users.Where(s=>uids.Contains(s.Uid)).ToListAsync();
> ```
> 
> The codes above works ok with 8.0.11. Exception was:
> 
> ```
>       An unhandled exception has occurred while executing the request.
>       System.InvalidOperationException: The LINQ expression '@__uids_0' could not be translated. Additional information: Primitive collections support has not been enabled. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
>          at Pomelo.EntityFrameworkCore.MySql.Query.Internal.MySqlQueryableMethodTranslatingExpressionVisitor.TranslatePrimitiveCollection(SqlExpression sqlExpression, IProperty property, String tableAlias)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateParameterQueryRoot(ParameterQueryRootExpression parameterQueryRootExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitExtension(Expression extensionExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.TranslateSubquery(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.<VisitMethodCall>g__TranslateAsSubquery|33_0(Expression expression, <>c__DisplayClass33_0&)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.CallBaseVisitMethodCall(MethodCallExpression methodCallExpression)
>          at Pomelo.EntityFrameworkCore.MySql.Query.ExpressionVisitors.Internal.MySqlSqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.Translate(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression, Boolean applyDefaultTypeMapping)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
>          at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
>          at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression[TResult](Expression query)
>          at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
>          at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass11_0`1.<ExecuteCore>b__0()
>          at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteCore[TResult](Expression query, Boolean async, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
>          at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
>          at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
>          at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
> ```

A solution is mentioned here:
#1960 

EnablePrimitiveCollectionsSupport() and TranslateParameterizedCollectionsToConstants()
was needed to fix it for me

**Comment by miroslavpopovic** [comment]
Thanks, @devnullue, I should have looked for existing issues first. Yes, just with `TranslateParameterizedCollectionsToConstants`, all our integration tests are passing.

And big thanks to @lauxjpn for all the hard work on this!

**Comment by lauxjpn** [comment]
> Hi, is there any estimates on when you would publish a stable EF Core 9.0 version? We've been following this issue since November.

@jossean-rivera Our [Schedule and Roadmap](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql#schedule-and-roadmap) is up-to-date, so we still plan to release the 9.0 GA version in January.

**Comment by CollinHerber** [comment]
We are nearing the end of January and my team is scheduled for .NET 9 upgrades for the next sprint (Start of February) from these initial release notes. Are we still on track for that or should I pivot?
We are just waiting on Pomelo GA at this time.

**Comment by rowanfuchs** [comment]
> We are nearing the end of January and my team is scheduled for .NET 9 upgrades for the next sprint (Start of February) from these initial release notes. Are we still on track for that or should I pivot? We are just waiting on Pomelo.EntityFrameworkCore GA at this time.

Have your team contribute to the package, that should definitely speed things up.

**Comment by raffidahmad** [comment]
I'm facing a weird issue, in version 8 the Column order annotation was working as expected, it still works but the following does not seem to work anymore for me
` [Column(Order = 99999)] [Timestamp] public DateTime UpdatedDate { get; set; }`
This does not seem to work anymore in migrations and is causing these columns to appear first even

**Comment by enisn** [comment]
![Image](https://github.com/user-attachments/assets/2a915a9f-7658-4313-9797-ff05be334467)

**Comment by mc2002tii** [comment]
@MeikelLP, my two cents is that it isn't the delays that are the problem.  It is the lack of communication about the delays.  And the same thing happens every year with this project.  (and I'm not trying to defend enisn's obnoxious cartoon; that doesn't help anything).

**Comment by zijianhuang** [comment]
I have been using the Pomelo.EntityFrameworkCore.MySql 9.0.0-preview.2.efcore.9.0.0 for many weeks in production.
1. As said by many above, "You can still use the EF Core 8 version which runs on the LTS .NET8 so you don't have to use EF Core 9.
You can use EF Core 8 in .NET 9 applications.", provided:
a. You have proper branching strategy.
b. You don't feel annoyed when package management of VS IDE keeps prompting you to grade to .NET 9 complaint components like EF Core 9.

I have comprehensive test suites for the DAL, so I am fairly confident omelo.EntityFrameworkCore.MySql 9.0.0-preview.2.efcore.9.0.0 won't break my app --- I won't say it won't break yours since my app is probably not complicated enough to encounter the edge cases you have experienced.

A trick when using 9.0.0-preview.2.efcore.9.0.0 is, you may need to explicitly reference Microsoft.EntityFrameworkCore.Relational or Microsoft.EntityFrameworkCore 9 or 9.01 in some of your csproj files when the compiler or runtime MIGHT complains downgrading of the package.

**Comment by MeikelLP** [comment]
>        b. You don't feel annoyed when package management of VS IDE keeps prompting you to grade to .NET 9 complaint 

@zijianhuang: You can use version ranges to prevent that: https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#version-ranges VS should not warn you then. Very useful tip I learned recently with FluentAssertions the fiasco

> [@MeikelLP](https://github.com/MeikelLP), my two cents is that it isn't the delays that are the problem. It is the lack of communication about the delays. And the same thing happens every year with this project. (and I'm not trying to defend enisn's obnoxious image; that doesn't help anything).

This is still a free (most likely hobbyish) project from some people from the internet. They are not required to implement anything. It's a blessing they do it for free. Oracle is worse even though they are paid. Also they are developer not PR guys. The are not required be transparent to respond at all. If you don't like them: Fork the project. That's the OSS spirit.

**Comment by boeschenstein** [comment]
Update Pomelo.EntityFrameworkCore.MySql from 8.0.2 to 9.0.0-preview.2.efcore.9.0.0

We have a "Contains" using an 'enum' in our query, which runs fine in version v8.0.2, but fails in v9:

```cs
public LoadDataSpecification(Guid mandatorId, List<ServiceType> serviceTypes)
    : base(f => f.MandatorId.Equals(mandatorId) && serviceTypes.Contains(f.ServiceType));
```

Without "Contains" this query runs fine in v9 as well:

```cs
public LoadDataSpecification(Guid mandatorId, List<ServiceType> serviceTypes)
    : base(f => f.MandatorId.Equals(mandatorId));
```

So can you please support "Contains" in v9, as well?

```txt
Result:
System.InvalidOperationException
HResult=0x80131509
Message=The LINQ expression '@__serviceTypes_1' could not be translated. 
Additional information: Primitive collections support has not been enabled. 
Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'.
See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
Source=Pomelo.EntityFrameworkCore.MySql
StackTrace:
at Pomelo.EntityFrameworkCore.MySql.Query.Internal.MySqlQueryableMethodTranslatingExpressionVisitor.TranslatePrimitiveCollection(SqlExpression sqlExpression, IProperty property, String tableAlias)
```

**Comment by miroslavpopovic** [comment]
> Update Pomelo.EntityFrameworkCore.MySql from 8.0.2 to 9.0.0-preview.2.efcore.9.0.0
> 
> We have a "Contains" using an 'enum' in our query, which runs fine in version v8.0.2, but fails in v9:
> 
> public LoadDataSpecification(Guid mandatorId, List<ServiceType> serviceTypes)
>     : base(f => f.MandatorId.Equals(mandatorId) && serviceTypes.Contains(f.ServiceType));
> 
> Without "Contains" this query runs fine in v9 as well:
> 
> public LoadDataSpecification(Guid mandatorId, List<ServiceType> serviceTypes)
>     : base(f => f.MandatorId.Equals(mandatorId));
> 
> So can you please support "Contains" in v9, as well?
> 
> Result:
> System.InvalidOperationException
> HResult=0x80131509
> Message=The LINQ expression '@__serviceTypes_1' could not be translated. 
> Additional information: Primitive collections support has not been enabled. 
> Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'.
> See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.
> Source=Pomelo.EntityFrameworkCore.MySql
> StackTrace:
> at Pomelo.EntityFrameworkCore.MySql.Query.Internal.MySqlQueryableMethodTranslatingExpressionVisitor.TranslatePrimitiveCollection(SqlExpression sqlExpression, IProperty property, String tableAlias)

@boeschenstein, this has been referenced in the past comments. I have asked the same question, and @devnullue has pointed me to https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1960. Adding option `.TranslateParameterizedCollectionsToConstants()` during MySQL configuration helped.

**Comment by Donistivanov** [comment]
Is anyone having trouble with seeding data the new way?

I am getting:

```
fail: Microsoft.EntityFrameworkCore.Database.Transaction[20205]
      An error occurred using a transaction.
An error occurred using a transaction.
EF is opening a connection to MySqlConnector.MySqlException (0x80004005): SAVEPOINT __EFSavePoint does not exist
   at MySqlConnector.Core.ServerSession.ReceiveReplyAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 1081
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 37
   at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 131
   at MySqlConnector.MySqlDataReader.InitAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 487
   at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 56
   at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 309
   at MySqlConnector.MySqlCommand.ExecuteNonQuery() in /_/src/MySqlConnector/MySqlCommand.cs:line 108
   at Microsoft.EntityFrameworkCore.Storage.RelationalTransaction.ReleaseSavepoint(String name)
```

**Comment by harry-hathorn** [comment]
Using version
9.0.0-preview.2.efcore.9.0.0

I have an error on expressions with an Include and Contains combined. Not sure if it Contains in general,

e.g. @miroslavpopovic  with a similar issue.

UserVariableType is an enum


` _context.Users
 .Include(x => x.UserVariables.Where(y => userVariables.Contains(y.UserVariableType)))
 .FirstOrDefaultAsync();`


System.InvalidOperationException: 'The LINQ expression '@__userVariables_0' could not be translated. Additional information: Primitive collections support has not been enabled. Either rewrite the query in a form that can be translated, or switch to client evaluation explicitly by inserting a call to 'AsEnumerable', 'AsAsyncEnumerable', 'ToList', or 'ToListAsync'. See https://go.microsoft.com/fwlink/?linkid=2101038 for more information.'

**Comment by JinHuihui1234** [comment]
When to publish the Release version？It's already February.

**Comment by srad** [comment]
> [@MeikelLP](https://github.com/MeikelLP), my two cents is that it isn't the delays that are the problem. It is the lack of communication about the delays. And the same thing happens every year with this project. (and I'm not trying to defend enisn's obnoxious image; that doesn't help anything).

I already mentioned that people (and I) would be willing to donate time. However, the project is under-documneted and not in a shape which would allow any outsider to help out or dig into the code with a reasonable amount of effort.

This project needs new direction in order to involve more people to help out. We can't just have two one or two people be blockers for such an important project.

**Comment by lauxjpn** [comment]
I have time on my hands again, probably until the end of summer. So we should be able to release the GA version in March.

**Comment by lauxjpn** [comment]
Prerelease [9.0.0-preview.3.efcore.9.0.0](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/9.0.0-preview.3.efcore.9.0.0) (compatible with EF Core `9.0.x`) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/9.0.0-preview.3.efcore.9.0.0).

It mainly contains the fix for the #1955 issue, so that using the latest preview does not change existing/expected query behavior.

**Comment by sommmen** [comment]
> Prerelease [9.0.0-preview.3.efcore.9.0.0](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/9.0.0-preview.3.efcore.9.0.0) (compatible with EF Core `9.0.x`) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/9.0.0-preview.3.efcore.9.0.0).
> 
> It mainly contains the fix for the [#1955](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1955) issue, so that using the latest preview does not change existing/expected query behavior.

Can i now remove the fixes suggested in that thread with this release?

**Comment by lauxjpn** [comment]
> Can i now remove the fixes suggested in that thread with this release?

@sommmen You can remove it if you want to (or just leave it in). Since PR #1971, we are setting the exact same option for you automatically, I you haven't set it yourself.

**Comment by featherbits** [comment]
> Is anyone having trouble with seeding data the new way (https://learn.microsoft.com/en-us/ef/core/modeling/data-seeding#configuration-options-useseeding-and-useasyncseeding-methods)?
> 
> I am getting:
> 
> ```
> fail: Microsoft.EntityFrameworkCore.Database.Transaction[20205]
>       An error occurred using a transaction.
> An error occurred using a transaction.
> MySqlConnector.MySqlException (0x80004005): SAVEPOINT __EFSavePoint does not exist
>    at MySqlConnector.Core.ServerSession.ReceiveReplyAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/ServerSession.cs:line 1081
>    at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in /_/src/MySqlConnector/Core/ResultSet.cs:line 37
>    at MySqlConnector.MySqlDataReader.ActivateResultSet(CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 131
>    at MySqlConnector.MySqlDataReader.InitAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, IDictionary`2 cachedProcedures, IMySqlCommand command, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlDataReader.cs:line 487
>    at MySqlConnector.Core.CommandExecutor.ExecuteReaderAsync(CommandListPosition commandListPosition, ICommandPayloadCreator payloadCreator, CommandBehavior behavior, Activity activity, IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/Core/CommandExecutor.cs:line 56
>    at MySqlConnector.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in /_/src/MySqlConnector/MySqlCommand.cs:line 309
>    at MySqlConnector.MySqlCommand.ExecuteNonQuery() in /_/src/MySqlConnector/MySqlCommand.cs:line 108
>    at Microsoft.EntityFrameworkCore.Storage.RelationalTransaction.ReleaseSavepoint(String name)
> ```
> 
> Troubleshooting steps: [dotnet/efcore#33319 (comment)](https://github.com/dotnet/efcore/issues/33319#issuecomment-2030777176)
> 
> Additional research: [mysql-net/MySqlConnector#1216](https://github.com/mysql-net/MySqlConnector/issues/1216)

`SAVEPOINT __EFSavePoint does not exist` error can be observed when entity gets saved with related entities. After error database shows that only a few related entities were inserted.

**Comment by lorinc-ambrozy-work** [comment]
> So we should be able to release the GA version in March.

Hi @lauxjpn , 
Do you already know when you'll have time to release the GA? I don’t want to rush you, but we’d like to plan ahead as well.

**Comment by MeikelLP** [comment]
> > So we should be able to release the GA version in March.
> 
> Hi [@lauxjpn](https://github.com/lauxjpn) , Do you already know when you'll have time to release the GA? I don’t want to rush you, but we’d like to plan ahead as well.

This is free and open source software. Do not plan your business based around free time and motivation of hobbyists.

If you really wanna plan ahead: Plan in some time to contribute to this project

**Comment by EmmanuelGonzalezTaveras** [comment]
I really like this repository and have great respect and admiration for the work you've done. However, given how fast the ecosystem is evolving and the delays in updates, we need to start considering alternatives before this project ends up like AutoMapper,  MassTransit or MediatR.

**Comment by MPITech** [comment]
>  we need to start considering alternatives

Who is stopping you?

**Comment by hoerup** [comment]
Could we at least get a preview4 release so we can test with latest fixes ?

**Comment by OliverRC** [comment]
EF Core 9 was released in November 2024 and we still don't have a stable GA version

**Comment by S-Muneeb-ALi** [comment]
@lauxjpn Meri toh shaadi bhi ho gayi, bachche bhi puchhte hain, 'Papa, yeh Open EF Core 9 kab release hoga?' Main kya jawab doon?

**Comment by dajma00** [comment]
Bhai, ye open source hai. Sabr karo.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by renemadsen** [comment]
@lauxjpn what are the blocking issue that needs to be fixed before it can become a GA for dotnet 9?

We are about to hit 6 months before dotnet 10 will be released and the cycle starts over, so it would be nice if we could make some kind of progress. 

I'm not commenting to just complain, but to know what needs to be done and maybe fix some of the issues, so we all are able to move forward :-)

**Comment by noismaster** [comment]
> NativeAOT and query precompilation are highly experimental features, and are not yet suited for production use.

If precompiled queries are "highly experimental" and not recommended for production, to they need to be in GA release?

Why not do a separate release later?

**Comment by Donistivanov** [comment]
In the great fields of cooperation, we have to remember to let people do what they think is best. Vote thumbs up!

**Comment by zijianhuang** [comment]
@boschin-git 
As mentioned in my comments above in this thread, I have been using the RC release since last December in a few commercial projects. As long as you have comprehensive test suites upon the DAL layer and the Web API level in both test and staging, you should be feeling comfortable using the RC component in production, as it is not uncommon some "production" releases of 3rd components are rather buggy, causing problems in my test suites.

**Comment by S-Muneeb-ALi** [comment]
@lauxjpn , woh Pomelo on EF Core 9 waali baat yaad hai? Ab toh hadd ho gayi! Main dada ban gaya hoon, aur ab mere pota-poti bhi mujhse puchhte hain, 'Dada ji, yeh Pomelo on EF 9 kab tak aayega?' EF Core 9 to last year hi aa gaya tha. Please yaar, koi update hai Pomelo ki taraf se? Mere budhape mein thoda sukoon mile!

**Comment by lajith111** [comment]
@lauxjpn  Thanks so much for the great work on this project! Just wondering if there’s any update planned soon. Looking forward to what’s next!

**Comment by naruto1227** [comment]
Hello, may I ask if it's available now? What's the plan for Net10?

**Comment by lajith111** [comment]
Any one used https://www.nuget.org/packages/MySql.EntityFrameworkCore .. Please  Suggestion need to upgrade my .net project

**Comment by holly-hacker** [comment]
> Any one used [nuget.org/packages/MySql.EntityFrameworkCore](https://www.nuget.org/packages/MySql.EntityFrameworkCore) .. Please Suggestion need to upgrade my .net project

It's worth noting that MySql.EntityFrameworkCore is licensed under a variant of GPL-2.0, so it's a non-started for many projects.

**Comment by adamtash** [comment]
We've been using Pomelo successfully in production for several years and we're grateful for the value it has provided our applications. Given the preview status of .NET 9.0 support and various other reasons, we've migrated to PostgreSQL with Npgsql EF Provider.

For others considering a similar path, I vibe coded a migration tool to migrate data from MySql to Postgres, since the popular PGLoader tool couldn't handle our complex schema and relationships out of the box: [https://github.com/adamtash/MySQL2Postgres](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)

**Comment by zijianhuang** [comment]
@adamtash 
One of the reasons why we would use Entity Framework (Core) is to enable us to switch among databases according to business content and contexts, and such flexibility gives us edge during development/testing and testing.
If a business app is well designed, its custom DAL should be independent of DB engines, and the switch to the other DB engine could be easily done through a plugin with respective DB client library along with config during deployment. 

Really, no migration needed from one DB engine to another.

For example, https://github.com/zijianhuang/AuthEF

**Comment by miroslavpopovic** [comment]
@zijianhuang, what @adamtash has created is a tool to migrate **data** from MySQL database to Postgres database, as existing tools didn't fulfill his needs. It is not intended for making any code changes.

**Comment by zijianhuang** [comment]
@miroslavpopovic , thank you for clarifying that.

@adamtash , as mentioned above, I have used the RC release of this lib since last December in a few complex business apps on .NET 9. I am curious what sort of errors you see in your .NET 9 apps utilizing Pomelo lib? though likely your DB schema could be much more complex than mine.

**Comment by rgwsk65** [comment]
@adamtash - We also decided to migrate to Postgres

**Comment by renemadsen** [comment]
I've created this fork https://github.com/microting/Pomelo.EntityFrameworkCore.MySql since we don't want to have any preview/rc packages in our production env.
It will not implement any new features or fix any bugs/performance issues.

The sole purpose is just to be up to date with .net and any other dependencies.
A nuget package can be found here https://www.nuget.org/packages/Microting.EntityFrameworkCore.MySql/ and is a complete dropin replacement, so any code changes should be to a minimum.

So it's not here to compete with the Pomelo, but we need to move forward and we are not in a position where it at all makes any sense to migrate to lets say Postgres.

**Comment by naruto1227** [comment]
Hello, What's the plan for efcore10?

**Comment by lauxjpn** [comment]
Release [9.0.0](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/releases/tag/9.0.0) is now available on [nuget.org](https://www.nuget.org/packages/Pomelo.EntityFrameworkCore.MySql/9.0.0).

We took our time with this one, so thanks everybody for being patient.
Also special thanks to everyone who contributed to this release!

This issue will continue to track `9.0.x` releases.
To follow our EF Core 10 targeted releases, track #2007.

**Comment by zijianhuang** [comment]
Great work, while I have been using the RC2 in production since last December. At least in my use cases of complex business apps, RC2 has been already on production quality.
Case closed.

**Comment by chendjouCaleb** [comment]
Thank you @lauxjpn for this 9.0.0 release.

**Comment by vanbukin** [comment]
@lauxjpn Thank you for release!

**Comment by khairulazmimatesa** [comment]
@lauxjpn Thank your for the great release!

**Comment by naruto1227** [comment]
Thank your
